kong/db/strategies/postgres/connector.lua

```diff
@@ -329,23 +329,22 @@ function _mt:init_worker(strategies)
       }
     end
 
-    cleanup_statement = concat({
-      "BEGIN;",
-      concat(cleanup_statement, "\n"),
-      "COMMIT;"
-    }, "\n")
-
     return timer_every(60, function(premature)
       if premature then
         return
       end
 
-      local ok, err = self:query(cleanup_statement)
-      if not ok then
-        if err then
-          log(WARN, "unable to clean expired rows from postgres database (", err, ")")
-        else
-          log(WARN, "unable to clean expired rows from postgres database")
+      for i, statement in ipairs(cleanup_statements) do
```

Comment:
I definitely agree with getting rid of the transaction for this operation, but I think that a unique query (or at least, a single round trip from Kong to PostgreSQL) would be better than sending many smaller queries, especially considering the lack of control over reused connections and additional network roundtrip this approach would cause regardless. Any reason for the individual queries?

Reply:
I thought about it, but reporting the table name was easier with multiple. I can look other ways.

Reply:
I tried different approaches, unfortunately, e.g. this:
```psql
DELETE FROM "oauth2_authorization_codes" WHERE "ttl" < CURRENT_TIMESTAMP AT TIME ZONE 'UTC';
DELETE FROM "oauth2_tokens" WHERE "ttl" < CURRENT_TIMESTAMP AT TIME ZONE 'UTC';
DELETE FROM "sessions" WHERE "ttl" < CURRENT_TIMESTAMP AT TIME ZONE 'UTC';
```

Would require `BEGIN EXCEPTION` handling on each statement which can be more expensive than it is beneficial. Also makes it hard to know and report what query errored. If for example the second will error, the third one will not be executed without proper exception handling.

So I think the current n-roundtrips is the best solution for this, unless someone has something better to offer?

Reply:
I wonder if there is anything to be done at the driver level for such use-cases (e.g. see https://github.com/leafo/pgmoon#result-err-partial-num_queries--postgresqueryquery_string and similar, although they don't do quite we want here as far as I can tell).

Nonetheless, we can keep the current n-roundtrips implementation which is already an improvement :+1:

Reply:
We could also run this query only when the schema has `ttl = true`, since we already assume that the TTL postgres column be present in the table when that is the case. This way we significantly reduce our margin for error and the logs noisiness. Actually, if we don't already do this, can we update this patch to do it?

Reply:
I tried that `partial` pgmoon thing, but problem is that it stops on first error. Then the later statements don't get executed. We could have some logic to try `single roundtrip` if not working (obviously should always work), revert to `n-queries`. The code already checks the `ttl=true` on schemas.

Reply:
Alright, the current approach sounds good enough then given the constraints of other approaches :+1: